### PR TITLE
feat: Refactor usernames and nicknames for a better system

### PR DESCRIPTION
**Pull request by kyranet**
This new system also helps querying the API less!

### CODE REVIEW

**Review by cfanoulis — DISMISSED** [review]
LGTM (at least the code I could understand)
minor nit I just thought about just as I left this: a helper to return the username#tag to skip the template literals?

File: src/arguments/membername.ts [hunk 1]
```diff
@@ -10,7 +10,7 @@ export default class extends Argument {
 		const resMember = await this.resolveMember(message, arg);
 		if (resMember) return resMember;
 
-		const result = await new FuzzySearch(message.guild!.memberUsernames, entry => entry, filter).run(message, arg, possible.min || undefined);
+		const result = await new FuzzySearch(message.guild!.nicknames.mapUsernames(), entry => entry, filter).run(message, arg, possible.min || undefined);
```
Comment (favna) [inline src/arguments/membername.ts:13]: I thought DJS fixed the need for the non null assertion for message.guild? Same for username if so.

File: src/commands/Tools/poll.ts [hunk 2]
```diff
@@ -141,10 +141,10 @@ export default class extends SkyraCommand {
 	public async _resolveUsers(message: KlasaMessage, users: string[]) {
 		const output: string[] = [];
 		for (const user of users) {
-			let resolved: string | undefined;
+			let resolved: string | null | undefined;
```
Comment (favna) [inline src/commands/Tools/poll.ts:144]: What's the point of `null | undefined` when you're not checking specifically `=== null` or `=== undefined` afterwards?

File: src/lib/util/FuzzySearch.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
-import { Collection, Message } from 'discord.js';
+import { Message } from 'discord.js';
 import { util } from 'klasa';
 import { levenshtein } from './External/levenshtein';
+import { Collection } from '@discordjs/collection';
```
Comment (favna) [inline src/lib/util/FuzzySearch.ts:4]: Why not import from discord.js?

**Review by favna — COMMENTED** [review]
Few more remarks

File: src/arguments/membername.ts (hunk 1 above)
Comment (kyranet) [reply src/arguments/membername.ts]: https://github.com/kyranet/Skyra/blob/4cb98c2aa110d0ef4ae2aa062920acf27705635d/src/commands/Tools/poll.ts#L147

Returns `string | null`.

File: src/commands/Tools/poll.ts (hunk 2 above)
Comment (kyranet) [reply src/commands/Tools/poll.ts:144]: https://github.com/kyranet/Skyra/blob/4cb98c2aa110d0ef4ae2aa062920acf27705635d/src/commands/Tools/poll.ts#L147

Returns `string | null`.

File: src/arguments/membername.ts (hunk 1 above)
Comment (kyranet) [reply src/arguments/membername.ts:13]: Messages sent in a DM channel have `guild === null`.

File: src/lib/util/FuzzySearch.ts (hunk 3 above)
Comment (kyranet) [reply src/lib/util/FuzzySearch.ts:4]: discord.js's Collection has `toJSON()`, which we don't need.

**Review by vladfrangu — CHANGES_REQUESTED** [review]
A thing I noticed

File: src/lib/util/util.ts [hunk 4]
```diff
@@ -294,6 +295,14 @@ export function getColor(message: Message) {
 	return message.author.settings.get(UserSettings.Color) || (message.member && message.member.displayColor) || BrandingColors.Primary;
 }
 
+const ROOT = 'https://cdn.discordapp.com';
```
Comment (vladfrangu) [inline src/lib/util/util.ts:298]: This is wrong. Iirc its discriminator % 5 

Good job FastHub... Discriminator line if the user has no avatar

Comment (kyranet) [reply src/lib/util/util.ts:298]: Fixed in c9c71c1

File: src/lib/util/FuzzySearch.ts (hunk 3 above)
Comment (favna) [reply src/lib/util/FuzzySearch.ts:4]: Yes but for consistency in the code I would think using the discord.js one would be better. Surely having that method and not even calling it has no real overhead?

Comment (kyranet) [reply src/lib/util/FuzzySearch.ts:4]: It's unnecessary, that's why.

And as a general rule, I avoid all of discord.js's `toJSON` at all costs.

It gives all the data serialized plus bonus useless stuff.

**Review by favna — APPROVED** [review]

**Pull request merged**